File: tests/parsers/lecture-parser.test.ts
import { describe, it, expect } from 'vitest';
import { parseLectures } from '../../src/parsers/lecture-parser';

describe('parseLectures', () => {
  it('should parse lectures from table layout with course grouping', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3>CS101 - Programming Fundamentals</h3>
  <table>
    <tr><th>Title</th><th>Type</th><th>Duration</th><th>Status</th></tr>
    <tr><td>Introduction</td><td>Video</td><td>45 min</td><td>Watched</td></tr>
    <tr><td>Variables</td><td>Video</td><td>50 min</td><td>New</td></tr>
  </table>
</body>
</html>`;

    const lectures = parseLectures(html);
    expect(lectures).toHaveLength(2);
    expect(lectures[0].courseCode).toBe('CS101');
    expect(lectures[0].title).toBe('Introduction');
    expect(lectures[0].type).toBe('Video');
    expect(lectures[0].duration).toBe('45 min');
    expect(lectures[0].status).toBe('watched');
    expect(lectures[1].status).toBe('new');
  });

  it('should parse unwatched status', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <table>
    <tr><th>Title</th><th>Status</th></tr>
    <tr><td>Functions</td><td>Unwatched</td></tr>
  </table>
</body>
</html>`;

    const lectures = parseLectures(html);
    expect(lectures).toHaveLength(1);
    expect(lectures[0].status).toBe('unwatched');
  });

  it('should parse new status by default when no status indicator', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <div class="lecture-card">
    <div class="title">Introduction</div>
  </div>
</body>
</html>`;

    const lectures = parseLectures(html);
    expect(lectures).toHaveLength(1);
    expect(lectures[0].status).toBe('new');
  });

  it('should parse lecture type and duration from card layout', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <h3>CS101 - Programming</h3>
  <div class="lecture-card">
    <div class="title">Lecture 1</div>
    <div class="type">Video</div>
    <div class="duration">45 min</div>
    <div class="status">Watched</div>
    <a href="/Video.aspx?id=1">Watch</a>
  </div>
</body>
</html>`;

    const lectures = parseLectures(html);
    expect(lectures).toHaveLength(1);
    expect(lectures[0].courseCode).toBe('CS101');
    expect(lectures[0].type).toBe('Video');
    expect(lectures[0].duration).toBe('45 min');
    expect(lectures[0].status).toBe('watched');
    expect(lectures[0].url).toBe('/Video.aspx?id=1');
  });

  it('should extract week number', () => {
    const html = `<!DOCTYPE html>
<html>
<body>
  <table>
    <tr><th>Title</th><th>Week</th></tr>
    <tr><td>Lecture 1</td><td>Week 1</td></tr>
  </table>
</body>
</html>`;

    const lectures = parseLectures(html);
    expect(lectures).toHaveLength(1);
    expect(lectures[0].week).toBe(1);
  });

  it('should return empty array when no lectures found', () => {
    expect(parseLectures('<html><body>No lectures</body></html>')).toEqual([]);
  });
});

File: vulms_dart/test/parsers/lecture_parser_test.dart
import 'dart:io';
import 'package:test/test.dart';
import 'package:vulms_dart/src/parsers/lecture_parser.dart';
import 'package:vulms_dart/src/models/lecture.dart';

void main() {
  final fixture = File('test/fixtures/lectures_table.html').readAsStringSync();
  final emptyFixture = File('test/fixtures/empty_page.html').readAsStringSync();

  group('LectureParser.parse()', () {
    group('table layout', () {
      test('parses 3 lectures', () {
        final result = LectureParser.parse(fixture);
        expect(result, hasLength(3));
      });

      test('extracts OOP Introduction details', () {
        final result = LectureParser.parse(fixture);
        final l1 = result.firstWhere(
          (l) => l.title.contains('OOP Introduction'),
        );
        expect(l1.courseCode, 'CS101');
        expect(l1.title, 'OOP Introduction');
        expect(l1.type, 'Video');
        expect(l1.duration, '45 min');
        expect(l1.week, 1);
      });

      test('extracts Data Types lecture details', () {
        final result = LectureParser.parse(fixture);
        final l2 = result.firstWhere(
          (l) => l.title.contains('Data Types'),
        );
        expect(l2.title, 'Data Types and Variables');
        expect(l2.type, 'Video');
        expect(l2.duration, '30 min');
        expect(l2.week, 1);
      });

      test('extracts Control Structures lecture details', () {
        final result = LectureParser.parse(fixture);
        final l3 = result.firstWhere(
          (l) => l.title.contains('Control'),
        );
        expect(l3.title, 'Control Structures');
        expect(l3.type, 'Video');
        expect(l3.duration, '55 min');
        expect(l3.week, 2);
      });
    });

    group('status determination', () {
      test('Watched maps to watched', () {
        final result = LectureParser.parse(fixture);
        final l1 = result.firstWhere(
          (l) => l.title.contains('OOP Introduction'),
        );
        expect(l1.status, LectureStatus.watched);
      });

      test('New maps to newLecture', () {
        final result = LectureParser.parse(fixture);
        final l2 = result.firstWhere(
          (l) => l.title.contains('Data Types'),
        );
        expect(l2.status, LectureStatus.newLecture);
      });

      test('Not yet viewed maps to unwatched', () {
        final result = LectureParser.parse(fixture);
        final l3 = result.firstWhere(
          (l) => l.title.contains('Control'),
        );
        expect(l3.status, LectureStatus.unwatched);
      });
    });

    group('week number extraction', () {
      test('extracts week number from "Week 1"', () {
        final result = LectureParser.parse(fixture);
        final l1 = result.firstWhere(
          (l) => l.title.contains('OOP Introduction'),
        );
        expect(l1.week, 1);
      });

      test('extracts week number from "Week 2"', () {
        final result = LectureParser.parse(fixture);
        final l3 = result.firstWhere(
          (l) => l.title.contains('Control'),
        );
        expect(l3.week, 2);
      });
    });

    group('deduplication', () {
      test('removes duplicate lectures', () {
        const html = '<html><body>'
            '<h3 class="m-subheader__title">CS101</h3>'
            '<table>'
            '<tr><th>Title</th><th>Type</th><th>Duration</th><th>Week</th><th>Status</th></tr>'
            '<tr><td>OOP Intro</td><td>Video</td><td>45 min</td><td>Week 1</td><td>Watched</td></tr>'
            '<tr><td>OOP Intro</td><td>Video</td><td>45 min</td><td>Week 1</td><td>Watched</td></tr>'
            '</table></body></html>';
        final result = LectureParser.parse(html);
        expect(result, hasLength(1));
      });
    });

    group('empty page', () {
      test('returns empty list for page with no lectures', () {
        final result = LectureParser.parse(emptyFixture);
        expect(result, isEmpty);
      });
    });
  });
}
